test_utils.go

```diff
@@ -247,6 +264,7 @@ func (a TestClient) LongCallResult(ctx context.Context) (ret int, err error) {
 
 type mockCodec struct {
 	elems []interface{}
+	mtx   sync.Mutex
```

Comment:
hmm, there are a few unprotected reads of md.elems in NumElements and Decode.

But I think mockCodec/blockingMockCodec can just be replaced by a thin wrapper around a channel! Something like:

```
type mockCodec struct {
  elems chan interface{}
}

func newMockCodec(blocking bool, elems ...interface{}) mockCodec {
  // make mc.elems with capacity(len(elems)) and fill it up with elems
}

func Decode() {
  if blocking {
    i := <-mc.elems
    ...
  } else {
    select {
      case <-mc.elems:
        // decode
      default:
        error
    }
  }
}

func Encode() {
  // send on channel
}
```

Then no need for locking, etc.